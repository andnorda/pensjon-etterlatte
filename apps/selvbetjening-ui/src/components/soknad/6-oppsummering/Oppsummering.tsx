import { useSoknadContext } from "../../../context/soknad/SoknadContext";
import { sendSoeknad } from "../../../api";
import { useHistory } from "react-router-dom";
import { Normaltekst, Systemtittel } from "nav-frontend-typografi";
import React, { memo, useState } from "react";
import { SkjemaGruppe } from "nav-frontend-skjema";
import SoknadSteg from "../../../typer/SoknadSteg";
import AlertStripe from "nav-frontend-alertstriper";
import Navigasjon from "../../felles/Navigasjon";
import ObjectTreeReader from "../../../utils/ObjectTreeReader";
import { IAvdoed, IOmBarn, ISoeker, ISoekerOgAvdoed } from "../../../typer/person";
import TekstGruppe from "./fragmenter/TekstGruppe";
import { v4 as uuid } from "uuid";
import { useTranslation } from "react-i18next";
import Ekspanderbartpanel from "nav-frontend-ekspanderbartpanel";
import { StegPath } from "../../../context/steg/steg";
import { EditFilled } from "@navikt/ds-icons";
import Lenke from "nav-frontend-lenker";
import { ISituasjon } from "../../../typer/situasjon";

const Oppsummering: SoknadSteg = memo(({ forrige }) => {
    const { t, i18n } = useTranslation();
    const history = useHistory();
    const { state } = useSoknadContext();

    const [senderSoeknad, setSenderSoeknad] = useState(false);
    const [error, setError] = useState(false);

    const send = () => {
        setSenderSoeknad(true);
        setError(false);

        sendSoeknad(state)
            .then((soknadId) => {
                history.push(`/soknad/sendt/${soknadId}`);
            })
            .catch((error) => {
                console.log(error);
                setSenderSoeknad(false);
                setError(true);
            });
    };

    const getBaseKey = (string: string) => {
        return string.replace(/(.\d+)/g, "");
    };

    const otr = new ObjectTreeReader(i18n);

    const omDeg = otr.traverse<ISoeker>(state.omDeg, "omDeg");
    const omDegOgAvdoed = otr.traverse<ISoekerOgAvdoed>(state.omDegOgAvdoed, "omDegOgAvdoed");
    const omDenAvdoede = otr.traverse<IAvdoed>(state.omDenAvdoede, "omDenAvdoede");
    const dinSituasjon = otr.traverse<ISituasjon>(state.dinSituasjon, "dinSituasjon");

    // TODO: Fikse m√•ten barn og tidligere arbeidsforhold vises.
    const opplysningerOmBarn = otr.traverse<IOmBarn>(state.opplysningerOmBarn, "omBarn");

    const ekspanderbartPanel = (tittel: string, tekster: any[], path: StegPath) => (
        <Ekspanderbartpanel tittel={tittel} className={"oppsummering"} apen={true}>
            {!tekster.length && (
                <SkjemaGruppe>
                    <Normaltekst>{t("felles.ingenInfo")}</Normaltekst>
                </SkjemaGruppe>
            )}

            {tekster.map(({ key, val }) => (
                <TekstGruppe key={uuid()} tittel={t(getBaseKey(key))} innhold={t(val)} />
            ))}

            <Lenke href={`/soknad/steg/${path}`} className={senderSoeknad ? "disabled" : ""}>
                <EditFilled />
                <span>{t("felles.endreSvar")}</span>
            </Lenke>
        </Ekspanderbartpanel>
    );

    return (
        <>
            <SkjemaGruppe>
                <Systemtittel className={"center"}>{t("oppsummering.tittel")}</Systemtittel>
            </SkjemaGruppe>

            <SkjemaGruppe>
                <Normaltekst>{t("oppsummering.beskrivelse")}</Normaltekst>
            </SkjemaGruppe>

            {ekspanderbartPanel(t("omDeg.tittel"), omDeg, StegPath.OmDeg)}

            {ekspanderbartPanel(t("omDegOgAvdoed.tittel"), omDegOgAvdoed, StegPath.OmDegOgAvdoed)}

            {ekspanderbartPanel(t("omDenAvdoede.tittel"), omDenAvdoede, StegPath.OmAvdoed)}

            {ekspanderbartPanel(t("dinSituasjon.tittel"), dinSituasjon, StegPath.DinSituasjon)}

            {ekspanderbartPanel(t("omBarn.tittel"), opplysningerOmBarn, StegPath.OmBarn)}

            <br />

            {error && (
                <SkjemaGruppe>
                    <AlertStripe type={"feil"}>{t("oppsummering.feilVedSending")}</AlertStripe>
                </SkjemaGruppe>
            )}

            <Navigasjon forrige={{ onClick: forrige }} send={{ onClick: send }} disabled={senderSoeknad} />
        </>
    );
});

export default Oppsummering;
