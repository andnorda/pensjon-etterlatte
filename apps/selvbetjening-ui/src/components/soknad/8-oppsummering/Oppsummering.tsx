import { useSoknadContext } from "../../../context/soknad/SoknadContext";
import { sendSoeknad } from "../../../api";
import { useHistory } from "react-router-dom";
import { Ingress, Normaltekst, Systemtittel } from "nav-frontend-typografi";
import React, { useState } from "react";
import { Fareknapp, Flatknapp, Hovedknapp, Knapp } from "nav-frontend-knapper";
import { SkjemaGruppe } from "nav-frontend-skjema";
import OppsummeringSoeker from "./fragmenter/OppsummeringSoeker";
import OppsummeringSituasjon from "./fragmenter/OppsummeringSituasjon";
import OppsummeringAvdoed from "./fragmenter/OppsummeringAvdoed";
import SoknadSteg from "../../../typer/SoknadSteg";
import { useTranslation } from "react-i18next";
import OppsummeringBarn from "./fragmenter/OppsummeringBarn";
import OppsummeringTidlArbeid from "./fragmenter/OppsummeringTidlArbeid";
import OppsummeringArbeidsforhold from "./fragmenter/OppsummeringArbeidsforhold";
import OppsummeringAndreYtelser from "./fragmenter/OppsummeringAndreYtelser";
import { default as Modal } from "nav-frontend-modal";
import { ActionTypes } from "../../../context/soknad/soknad";

const Oppsummering: SoknadSteg = ({ forrige }) => {
    const history = useHistory();
    const { t } = useTranslation();
    const { state, dispatch } = useSoknadContext();

    const {
        situasjon,
        opplysningerOmSoekeren,
        opplysningerOmDenAvdoede,
        opplysningerOmBarn,
        tidligereArbeidsforhold,
        naavaerendeArbeidsforhold,
        andreYtelser
    } = state;

    const [isOpen, setIsOpen] = useState<boolean>(false);

    const send = () => {
        sendSoeknad(state)
            .then((r) => {
                console.log(r);

                history.push("/soknad/sendt");
            })
            .catch((error) => {
                // TODO: Håndtere feil. Redirect til feilside?
                console.error(error);


            });
    };

    const avbrytSoeknad = () => {
        dispatch({ type: ActionTypes.TILBAKESTILL })

        window.location.href = "https://www.nav.no"
    }

    return (
        <>
            <SkjemaGruppe>
                <Systemtittel className={"center"}>Oppsummering</Systemtittel>
            </SkjemaGruppe>

            <SkjemaGruppe>
                <Normaltekst>Les gjennom oppsummeringen av din søknad før du sender.</Normaltekst>
                <Normaltekst>Hvis du trenger å gjøre endringer, kan du gå tilbake og gjøre det. </Normaltekst>
            </SkjemaGruppe>

            <OppsummeringSituasjon state={situasjon!!} />

            <OppsummeringSoeker state={opplysningerOmSoekeren!!} />

            <OppsummeringAvdoed state={opplysningerOmDenAvdoede!!} />

            <OppsummeringBarn state={opplysningerOmBarn} />

            <OppsummeringTidlArbeid state={tidligereArbeidsforhold} />

            <OppsummeringArbeidsforhold state={naavaerendeArbeidsforhold!!} />

            <OppsummeringAndreYtelser state={andreYtelser!!} />

            <br />

            <Modal
                isOpen={isOpen}
                onRequestClose={() => setIsOpen(false)}
                closeButton={true}
                contentLabel={"aasdfsdf"}
            >
                <SkjemaGruppe>
                    <Ingress>
                        Er du helt sikker på at du vil avbryte søknaden?
                    </Ingress>
                </SkjemaGruppe>

                <SkjemaGruppe>
                    <Hovedknapp htmlType={"button"} onClick={() => setIsOpen(false)}>
                        Nei, jeg vil fortsette søknaden
                    </Hovedknapp>
                </SkjemaGruppe>

                <SkjemaGruppe>
                    <Fareknapp htmlType={"button"} onClick={avbrytSoeknad}>
                        Ja, avbryt søknaden
                    </Fareknapp>
                </SkjemaGruppe>
            </Modal>

            <SkjemaGruppe className={"navigasjon-rad"}>
                <Knapp htmlType={"button"} onClick={forrige}>
                    {t("knapp.tilbake")}
                </Knapp>

                <Hovedknapp htmlType={"button"} onClick={send}>
                    {t("knapp.sendSoeknad")}
                </Hovedknapp>
            </SkjemaGruppe>

            <SkjemaGruppe className={"navigasjon-rad"}>
                <Flatknapp htmlType={"button"} onClick={() => setIsOpen(true)}>
                    Avbryt
                </Flatknapp>
            </SkjemaGruppe>
        </>
    );
};

export default Oppsummering;
