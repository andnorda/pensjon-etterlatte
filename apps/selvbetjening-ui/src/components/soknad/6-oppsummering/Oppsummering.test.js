import { render } from "@testing-library/react";
import * as JSutils from "nav-frontend-js-utils";
import Oppsummering from "./Oppsummering";

jest.mock("react-i18next", () => ({
    // this mock makes sure any components using the translate hook can use it without a warning being shown
    useTranslation: () => {
        return {
            t: (str) => str,
            i18n: {
                changeLanguage: () => new Promise(() => {}),
            },
        };
    },
}));

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useLocation: jest.fn(() => ({ pathname: "/soknad/steg/oppsummering" })),
}));

JSutils.guid = jest.fn(() => "123");
describe("Om den avdÃ¸de", () => {
    it("Snapshot", () => {
        const { container } = render(<Oppsummering />);
        expect(container).toMatchSnapshot();
    });
});
