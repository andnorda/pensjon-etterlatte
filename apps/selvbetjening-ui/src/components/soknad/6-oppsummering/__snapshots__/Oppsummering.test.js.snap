// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Om den avd√∏de Snapshot 1`] = `
<div>
  <fieldset
    aria-invalid="false"
    class="skjemagruppe"
  >
    <h1
      class="center navds-title navds-title--m"
    >
      oppsummering.tittel
    </h1>
    <div
      aria-live="polite"
    />
  </fieldset>
  <fieldset
    aria-invalid="false"
    class="skjemagruppe"
  >
    <p
      class="navds-body-long"
    >
      oppsummering.beskrivelse
    </p>
    <div
      aria-live="polite"
    />
  </fieldset>
  <br />
  <fieldset
    aria-invalid="false"
    class="skjemagruppe navigasjon-footer"
  >
    <div
      class="navigasjon-rad"
    >
      <button
        class="navds-button navds-button--primary navds-button--m navds-body-short"
        type="button"
      >
        knapp.tilbake
      </button>
      <button
        class="navds-button navds-button--action navds-button--m navds-body-short"
        type="button"
      >
        knapp.sendSoeknad
         
      </button>
    </div>
    <div
      aria-live="polite"
    />
  </fieldset>
  <fieldset
    aria-invalid="false"
    class="skjemagruppe navigasjon-rad"
  >
    <button
      class="navds-button navds-button--secondary navds-button--m navds-body-short"
      type="button"
    >
      knapp.avbryt
    </button>
    <div
      aria-live="polite"
    />
  </fieldset>
</div>
`;
